src/core/experimental/refcount.d

```diff
@@ -0,0 +1,545 @@
+module core.experimental.refcount;
+
+private struct StatsAllocator
+{
+    version(CoreUnittest) size_t bytesUsed;
+
+    @trusted @nogc nothrow pure
+    void* allocate(size_t bytes) shared
+    {
+        import core.memory : pureMalloc;
+        if (!bytes) return null;
+
+        auto p = pureMalloc(bytes);
+        if (p is null) return null;
+        enum alignment = size_t.sizeof;
+        assert(cast(size_t) p % alignment == 0);
+
+        version (CoreUnittest)
+        {
+            static if (is(typeof(this) == shared))
+            {
+                import core.atomic : atomicOp;
+                atomicOp!"+="(bytesUsed, bytes);
+            }
+            else
+            {
+                bytesUsed += bytes;
+            }
+        }
+        return p;
+    }
+
+    @system @nogc nothrow pure
+    bool deallocate(void[] b) shared
+    {
+        import core.memory : pureFree;
+        assert(b !is null);
+
+        version (CoreUnittest)
+        {
+            static if (is(typeof(this) == shared))
+            {
+                import core.atomic : atomicOp;
+                assert(atomicOp!">="(bytesUsed, b.length));
+                atomicOp!"-="(bytesUsed, b.length);
+            }
+            else
+            {
+                assert(bytesUsed >= b.length);
+                bytesUsed -= b.length;
+            }
+        }
+        pureFree(b.ptr);
+        return true;
+    }
+
+    static shared StatsAllocator instance;
+}
+
+version (CoreUnittest)
+{
+    private shared StatsAllocator allocator;
+
+    private @nogc nothrow pure @trusted
+    void* pureAllocate(size_t n)
+    {
+        return (cast(void* function(size_t) @nogc nothrow pure)(&_allocate))(n);
+    }
+
+    private @nogc nothrow @safe
+    void* _allocate(size_t n)
+    {
+        return allocator.allocate(n);
+    }
+
+    private @nogc nothrow pure
+    void* pureDeallocate(T)(T[] b)
+    {
+        return (cast(void* function(T[]) @nogc nothrow pure)(&_deallocate!(T)))(b);
+    }
+
+    private @nogc nothrow
+    void* _deallocate(T)(T[] b)
+    {
+        allocator.deallocate(b);
+        return null;
+    }
+}
+
+struct __RefCount
+{
+    version(CoreUnittest) {} else
+    {
+        import core.memory : pureMalloc, pureFree;
```

Comment:
So I don't have any top level imports. I had pay as you go (for the user) in mind, but since `__RefCount` is not a template `core.memory` will be opened. If in the future the compiler will be able to avoid opening the internal imports if the type is not used, then this will behave nicely without changes.

Reply:
I think lazy importing will more likely happen for static or selective imports.